Add Order Instance For Header and Headers

isomarcte opened the pull request:
I left the old `Eq` based symbols in place so as not to break binary compatibility.

rossabaker commented on core/src/main/scala/org/http4s/Header.scala:
```diff
@@ -132,7 +132,16 @@ object Header {
-  implicit val HeaderEq: Eq[Header] = Eq.instance[Header] { (a, b) =>
-    a.name === b.name && a.value === b.value
-  }
+  @deprecated(message = "Please use HeaderOrder instead", since = "1.0.0-M8")
```
We can do this in series/0.21.  I can cherry-pick it if you don't want to rebase it, but we should change the version number.

isomarcte replied:
@rossabaker I'll rebase.

rossabaker commented on core/src/main/scala/org/http4s/Header.scala:
```diff
@@ -132,7 +132,16 @@ object Header {
-    a.name === b.name && a.value === b.value
-  }
+  @deprecated(message = "Please use HeaderOrder instead", since = "1.0.0-M8")
+  lazy val HeaderEq: Eq[Header] = HeaderOrder
```
Nit: since it's just an alias, a `def` is faster and lighter.

isomarcte replied:
I'll update this.

rossabaker reviewed (changes requested):
Nice! Can we also add `OrderTests` for these to make sure they're lawful?  I think they are.

isomarcte commented:
@rossabaker updated. 

* Aliases use `def` now. 
* Rebased and retargeted for `series/0.21`. 
* Updated the deprecation warning.
* Added `OrderTests` for `Header` and `Headers`.

rossabaker reviewed (approved):
Great. :+1: on green

isomarcte commented:
@rossabaker some tests appear to be failing, but I don't think they are related to this PR.

rossabaker commented:
Agreed. Unfortunately, the bigger the build gets, the more often this happens. I hope the munit work reducing unsafeRunSync()s helps.
